lib-api/src/main/java/org/ergoplatform/appkit/babelfee/BabelFeeOperations.java

```diff
@@ -0,0 +1,259 @@
+            .outputs(outBoxBuilder.build())
+            .build();
+
+    }
+
+    /**
+     * Tries to fetch a babel fee box from blockchain data source with the given unspent boxes loader
+     *
+     * @param ctx       current blockchain context
+     * @param loader    loader to receive unspent boxes
+     * @param tokenId   tokenId offered to swap
+     * @param feeAmount nanoerg amount needed to swap
+     * @return babel fee box satisfying the needs, or null if none available
+     */
+    @Nullable
+    public static InputBox findBabelFeeBox(BlockchainContext ctx, BoxOperations.IUnspentBoxesLoader loader, ErgoId tokenId, long feeAmount) {
+        ErgoContract contractForToken = new BabelFeeBoxContract().getContractForToken(tokenId, ctx.getNetworkType());
+        Address address = contractForToken.toAddress();
+        loader.prepare(ctx, Collections.singletonList(address), feeAmount, new ArrayList<>());
+
+        int page = 0;
+        List<InputBox> inputBoxes = loader.loadBoxesPage(ctx, address, 0);
+
+        InputBox returnBox = null;
+        long pricePerToken = Long.MAX_VALUE;
+
+        while (!inputBoxes.isEmpty() && returnBox == null) {
+
+            // find the cheapest box satisfying our fee amount needs
+            for (InputBox inputBox : inputBoxes) {
+                try {
+                    BabelFeeBoxState babelFeeBoxState = new BabelFeeBoxState(inputBox);
+                    if (babelFeeBoxState.getValueAvailableToBuy() > feeAmount && babelFeeBoxState.getPricePerToken() < pricePerToken)
+                        returnBox = inputBox;
+                } catch (Throwable t) {
+                    // ignore, check next
+                }
+            }
+
+            page++;
+            // get another page
+            inputBoxes = loader.loadBoxesPage(ctx, address, page);
+        }
+
+        return returnBox;
+    }
+
+    /**
+     * Creates a transaction builder for the given babel fee box and the amount to be covered by
+     * babel fees. The returned transaction builder can be used like normal and automatically will
+     * add the babel fee inbox and outbox to the transaction. You need to make sure that other
+     * inboxes cover the amount of tokens needed.
+     *
+     * @param txB            transaction builder the new transaction builder uses under the hoods
+     * @param babelBox       input babel box to make the swap with
+     * @param nanoErgToCover nanoergs to be covered by babel box, usually the fee amount needed, maybe a change amount as well
+     * @return transaction builder to be used
+     */
+    public static UnsignedTransactionBuilder getBabelFeeTransactionBuilder(
+        UnsignedTransactionBuilder txB,
+        InputBox babelBox,
+        long nanoErgToCover
+    ) {
+        return new BabelFeeTransactionBuilder(txB, babelBox, nanoErgToCover);
+    }
+
+    private static class BabelFeeTransactionBuilder implements UnsignedTransactionBuilder {
```

Comment:
Java doesn't support kotlin's extension by delegation pattern. And this creates a lot of boilerplate code. I think we better have the only UnsignedTransactionBuilder and all application protocols are build on top of it. This will give some uniformity to the codebase.

Reply:
So what do you suggest? Adding a `withBabelFeeBox` method to the general `UnsignedTransactionBuilder` interface which all transaction builders implementing the interface have to fullfil is too much in my opinion. Extending the transaction builder is not possible as clients cannot instantiaate an own version of transaction builder, the implementation used is defined by blockchain context. The last method that I see is providing a method to BlockchainOperations that takes inputs and outputs and a transaction builder to do the necessary calls on the builder, but this is very unnatural for client code.

I don't agree that the builder has any access to internals; it delegates all calls and changes the parameters for the ones that need to be changed.

Reply:
With new methods on builder this BabelFeeTransactionBuilder can be removed I guess?

Reply:
already removed :)